### DISCUSSION TITLE
How to accelerate simulations

### DISCUSSION THREAD

**Discussion by edorachlee** [rsachetto/MonoAlg3D_C#73] [General]
I thought this was worth asking as a new discussion. So far I have been having fun using MonoAlg3D but now I am thinking about how to run hundreds of simulations in a reasonable timeframe.

I am curious to hear what people's experiences are in balancing CPU vs GPU usage. If I have an ok GPU (right now I am able to use an NVIDIA A100), how will the number of CPU cores affect the speed? I am running with 12 cores for now. Given the same GPU, will I see dramatic improvement by increasing the number of cores?

Also, what are some configuration options to speed up the simulation? So far as I can see:
- [main] use_adaptivity = true
- [linear_system_solver] use_gpu = yes
- [ode_solver] use_gpu = true

If there are other options I can tweak I would appreciate your insights about them.

Best,
Lucas

**Comment by rsachetto** [comment]
The number of cores helps a little bit, but the hard work is done by the GPU. So I don´t think you will have a dramatic improvement by only increasing the number of cores.

The use_adaptivity option does not work when we are solving the linear system with the GPU. So, is usuallly faster to use the GPU to solve the linear system and disable adaptiviy.

As far as I know, the best combination of parameters is:

[main] use_adaptivity = false
[linear_system_solver] use_gpu = yes
[ode_solver] use_gpu = true

The ode solver speed will depend on the cellular model. More complicated models tend to gain more speedup when using the GPU. For simpler models,  it could be better to solve them using the CPU (they will be solved using multiple threads).

Another thing that impact the speed is how often will are saving the mesh (print_rate option). I also recommed to use the ensight format to save your simulations, as it is faster and use less disk than the VTU format.

If you have multiple GPUs you can launch multiple simulations by changing the gpu_id parameter (the first id is 0).

**Reply by edorachlee** [reply]
Thanks. So just to confirm, I will not be able to use the space- and time-adaptive techniques if I use the GPU? I am going off the abstract of your paper below.

Oliveira RS, Rocha BM, Burgarelli D, Meira Jr W, Constantinides C, dos Santos RW. Performance evaluation of GPU parallelization, space‐time adaptive algorithms, and their combination for simulating cardiac electrophysiology. Int J Numer Meth Biomed Engng. 2018;34:e2913. https://doi.org/10.1002/cnm.2913

Also, any considerations as to optimal RAM or GPU VRAM we need?

**Reply by rsachetto** [reply]
You can use the ode solver that is time adaptive and uses the GPU. The linear system solver is not compatible with the space adaptive technic, as it is slower.

For ram and GPU VRAM it depends on the mesh size and discretization. Usually, 32 gb of ram and 12 gb of GPU VRAM is more than enough.
